openedx/features/effort_estimation/block_transformers.py

```diff
@@ -0,0 +1,225 @@
+            # Rather than attempt to work around it, just set a note for ourselves to not show durations for this
+            # course at all. Better no estimate than a misleading estimate.
+            block_structure.set_transformer_data(cls, cls.DISABLE_ESTIMATION, True)
+
+    @classmethod
+    def _collect_html_effort(cls, block_structure, block_key, xblock, _cache):
+        """Records a word count for later reading speed calculations."""
+        try:
+            text = lxml.html.fromstring(xblock.data).text_content() if xblock.data else ''
+        except Exception as exc:  # pylint: disable=broad-except
+            raise cls.MissingEstimationData() from exc
+
+        block_structure.set_transformer_block_field(block_key, cls, cls.HTML_WORD_COUNT, len(text.split()))
+
+    @classmethod
+    def _collect_video_effort(cls, block_structure, block_key, xblock, cache):
+        """Records a duration for later viewing speed calculations."""
+        # Lookup all course video metadata at once rather than piecemeal, for performance reasons
+        if cls.CACHE_VIDEO_DURATIONS not in cache:
+            all_videos, _ = get_videos_for_course(str(block_structure.root_block_usage_key.course_key))
+            cache[cls.CACHE_VIDEO_DURATIONS] = {v['edx_video_id']: v['duration'] for v in all_videos}
+
+        # Check if we have a duration. If not, raise an exception that will stop this transformer from affecting
+        # this course.
+        duration = max(0, cache[cls.CACHE_VIDEO_DURATIONS].get(xblock.edx_video_id, 0))
+        if not duration:
+            raise cls.MissingEstimationData()
+
+        block_structure.set_transformer_block_field(block_key, cls, cls.VIDEO_DURATION, duration)
+
+        # Some videos will suggest specific start & end times, rather than the whole video. Note that this is only
+        # supported in some clients (other clients - like the mobile app - will play the whole video anyway). So we
+        # record this duration separately, to use instead of the whole video duration if the client supports it.
+        clip_duration = (xblock.end_time - xblock.start_time).total_seconds()
+        if clip_duration > 0:
+            block_structure.set_transformer_block_field(block_key, cls, cls.VIDEO_CLIP_DURATION, clip_duration)
+
+    def transform(self, usage_info, block_structure):
+        # Skip any transformation if our collection phase said to
+        cls = EffortEstimationTransformer
+        if block_structure.get_transformer_data(cls, cls.DISABLE_ESTIMATION, default=False):
+            return
+
+        # Get bucket for this experiment. 0 is no estimate. 1 is only on sections. 2 is only on subsections.
+        # If we drop the experiment but keep the feature - we should leave estimates on all blocks, and just have the
+        # UI show the estimate at the level they care about.
+        bucket = EFFORT_ESTIMATION_LOCATION_FLAG.get_bucket(course_key=usage_info.course_key)
```

Comment:
are we bucketing on the course and not on the user?  
not sure if we have a test plan written up, but I would have expected the bucketing to me on a user level with randomly assigned buckets

Reply:
We always bucket by user. There's no way not to. Waffle flag code (and our experiment flag code built on top of it) goes and gets the current request to get the user under consideration (this is one reason why we reset `request.user` when masquerading).

But the `course_key=` bit means we *also* bucket per course (that is, per-enrollment, rather than per-user). So rather than being bucketed into one experience across edx.org, you may see different experiences in each of your courses. I asked Tim which way he wanted it.

Reply:
ah ok, when i saw the course_key argument i misunderstood, thanks for clarifying!